src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -56,10 +56,29 @@ def ideal.is_homogeneous : Prop :=
 ∀ (i : ι) ⦃r : A⦄, r ∈ I → (graded_algebra.decompose 𝒜 r i : A) ∈ I
 
 /-- For any `semiring A`, we collect the homogeneous ideals of `A` into a type. -/
-abbreviation homogeneous_ideal : Type* := { I : ideal A // I.is_homogeneous 𝒜 }
+structure homogeneous_ideal extends submodule A A :=
+(is_homogeneous' : ideal.is_homogeneous 𝒜 to_submodule)
 
-instance : has_mem A (homogeneous_ideal 𝒜) :=
-{ mem := λ r I, r ∈ (I : ideal A) }
+variable {𝒜}
+/--Converting a homogeneous ideal to an ideal-/
+def homogeneous_ideal.to_ideal (I : homogeneous_ideal 𝒜) : ideal A := I.to_submodule
+
+lemma homogeneous_ideal.is_homogeneous (I : homogeneous_ideal 𝒜) :
+  I.to_ideal.is_homogeneous 𝒜 := I.is_homogeneous'
+
+lemma homogeneous_ideal.to_ideal_injective :
+  function.injective (homogeneous_ideal.to_ideal : homogeneous_ideal 𝒜 → ideal A) :=
+λ ⟨x, hx⟩ ⟨y, hy⟩ (h : x = y), by simp [h]
+
+instance homogeneous_ideal.set_like : set_like (homogeneous_ideal 𝒜) A :=
+{ coe := λ I, I.to_ideal,
+  coe_injective' := λ I J h, homogeneous_ideal.to_ideal_injective $ set_like.coe_injective h }
+
+@[ext] lemma homogeneous_ideal.ext {I J : homogeneous_ideal 𝒜}
+  (h : I.to_ideal = J.to_ideal) : I = J := homogeneous_ideal.to_ideal_injective h
+
+lemma homogeneous_ideal.mem_iff {I : homogeneous_ideal 𝒜} {x : A} : x ∈ I ↔ x ∈ I.to_ideal :=
```

Comment:
Should this swap directions and become a simp-lemma?

Reply:
done.

Reply:
done